### ISSUE TITLE
OIDC (bearer) middleware is not working in version 1.11.2.  

### ISSUE THREAD

**Issue by ben-kotvis** [dapr/components-contrib#3025]
## Expected Behavior

When I spin up a new cluster in Kubernetes or in a dev environment and add a bearer middleware to a configuration I expect to get a successful response if I check the JWT in (JWT.io) to make sure the issuer and audience are correct and they are correct.  

## Actual Behavior

I get a 401 all the time no matter what.  

## Steps to Reproduce the Problem

1. Spin up a new environment (1.11.2) and use bearer middleware on an app
2. Call the endpoint with a valid token
3. You will always get a 401

If you use version 1.9.0 it works.  Also, 1.10.0 works as well.  So it was introduced somewhere between 1.10.0 and 1.11.2.

**Comment by ItalyPaleAle** [comment]
Can you  share the Component YAML you are using please?

**Comment by ben-kotvis** [comment]
> Can you share the Component YAML you are using please?

```
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: bearertoken
spec:
  type: middleware.http.bearer
  version: v1
  metadata:
  - name: authority
    value: "[guid removed]" 
  - name: issuer
    value: "https://login.microsoftonline.com/[guid removed]/v2.0"
```


Here is the config:
```
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: appconfig
spec:
  httpPipeline:
    handlers:
    - name: bearertoken
      type: middleware.http.bearer
```

**Comment by ItalyPaleAle** [comment]
@ben-kotvis Thanks.

And could you please share the logs from Dapr? If possible, can you start Dapr with `--log-level=debug`?

**Comment by ben-kotvis** [comment]
> @ben-kotvis Thanks.
> 
> And could you please share the logs from Dapr? If possible, can you start Dapr with `--log-level=debug`?

Sure, see below: I actually had a typo above, but I corrected it from authority to audience.

```
> Executing command: daprd --app-id "proxy" --app-port "8080" --components-path "./proxy/components" --config "./proxy/components/config.yaml" --dapr-grpc-port "50008" --dapr-http-port "3500" "--enable-api-logging=true" "--enable-metrics=true" "--enable-profiling=true" --log-level "debug" --metrics-port "9091" --placement-host-address "localhost:50005" --dapr-http-max-request-size 2048 <

time="2023-08-02T16:09:04.737603227Z" level=info msg="starting Dapr Runtime -- version 1.11.2 -- commit 83ca1abb11ffe34211db55dcd36d96b94252827a" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.737627282Z" level=info msg="log level set to: debug" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.737723128Z" level=info msg="metrics server started on :9091/" app_id=proxy instance=d5ec20bb80c3 scope=dapr.metrics type=log ver=1.11.2
time="2023-08-02T16:09:04.73802002Z" level=debug msg="Loading config from file: ./proxy/components/config.yaml" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.738515607Z" level=debug msg="Found 0 resiliency configurations in resources path" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.738543582Z" level=info msg="Resiliency configuration loaded" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.738679536Z" level=info msg="standalone mode configured" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.738695649Z" level=info msg="app id: proxy" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.738703043Z" level=info msg="mTLS is disabled. Skipping certificate request and tls validation" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.738740612Z" level=info msg="Dapr trace sampler initialized: DaprTraceSampler(P=0.000100)" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.738940045Z" level=info msg="local service entry announced: proxy -> 172.17.0.3:48277" app_id=proxy component="mdns (nameResolution/v1)" instance=d5ec20bb80c3 scope=dapr.contrib type=log ver=1.11.2
time="2023-08-02T16:09:04.738959455Z" level=info msg="Initialized name resolution to mdns" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.738976392Z" level=info msg="Loading components…" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.739376213Z" level=debug msg="Found component: bearertoken (middleware.http.bearer/v1)" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.739393792Z" level=info msg="Waiting for all outstanding components to be processed" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.739442903Z" level=debug msg="Loading component: bearertoken (middleware.http.bearer/v1)" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.739472462Z" level=info msg="Component loaded: bearertoken (middleware.http.bearer/v1)" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.73949552Z" level=info msg="All outstanding components processed" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:04.739548862Z" level=debug msg="Fetching OpenID Configuration: https://login.microsoftonline.com/[removed]/v2.0/.well-known/openid-configuration" app_id=proxy component="bearertoken (middleware.http.bearer)" instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.092523557Z" level=debug msg="Found JWKS URL: https://login.microsoftonline.com/[removed]/discovery/v2.0/keys" app_id=proxy component="bearertoken (middleware.http.bearer)" instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.188506136Z" level=info msg="enabled middleware.http.bearer/http  middleware" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.188538092Z" level=info msg="Loading endpoints" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.188909766Z" level=info msg="Waiting for all outstanding http endpoints to be processed" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.188937389Z" level=info msg="All outstanding http endpoints processed" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.188958908Z" level=info msg="gRPC proxy enabled" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.189031858Z" level=info msg="gRPC server listening on TCP address: :50008" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.grpc.api type=log ver=1.11.2
time="2023-08-02T16:09:05.18905595Z" level=info msg="Enabled gRPC tracing middleware" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.grpc.api type=log ver=1.11.2
time="2023-08-02T16:09:05.189076671Z" level=info msg="Enabled gRPC metrics middleware" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.grpc.api type=log ver=1.11.2
time="2023-08-02T16:09:05.189183124Z" level=info msg="Registering workflow engine for gRPC endpoint: [::]:50008" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.grpc.api type=log ver=1.11.2
time="2023-08-02T16:09:05.189226256Z" level=info msg="API gRPC server is running on port 50008" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.189690984Z" level=info msg="enabled metrics http middleware" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.http type=log ver=1.11.2
time="2023-08-02T16:09:05.18971485Z" level=info msg="enabled tracing http middleware" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.http type=log ver=1.11.2
time="2023-08-02T16:09:05.189755178Z" level=info msg="HTTP server listening on TCP address: :3500" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.http type=log ver=1.11.2
time="2023-08-02T16:09:05.189814709Z" level=info msg="HTTP profiling server listening on: :7777" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.http type=log ver=1.11.2
time="2023-08-02T16:09:05.189846459Z" level=info msg="http server is running on port 3500" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.189862459Z" level=info msg="The request body size parameter is: 2048" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.189901473Z" level=info msg="gRPC server listening on TCP address: :48277" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.grpc.internal type=log ver=1.11.2
time="2023-08-02T16:09:05.189935489Z" level=info msg="Enabled gRPC tracing middleware" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.grpc.internal type=log ver=1.11.2
time="2023-08-02T16:09:05.189951721Z" level=info msg="Enabled gRPC metrics middleware" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.grpc.internal type=log ver=1.11.2
time="2023-08-02T16:09:05.189999108Z" level=info msg="internal gRPC server is running on port 48277" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.190019233Z" level=info msg="application protocol: http. waiting on port 8080.  This will block until the app is listening on that port." app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.190207436Z" level=info msg="application discovered on port 8080" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.190869634Z" level=info msg="Application configuration loaded" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.190897242Z" level=info msg="actors: state store is not configured - this is okay for clients but services with hosted actors will fail to initialize!" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.19096599Z" level=info msg="actor runtime started. actor idle timeout: 1h0m0s. actor scan interval: 30s" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.actor type=log ver=1.11.2
time="2023-08-02T16:09:05.190992404Z" level=info msg="Configuring workflow engine with actors backend" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.wfengine type=log ver=1.11.2
time="2023-08-02T16:09:05.191009109Z" level=info msg="Registering component for dapr workflow engine..." app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.191034802Z" level=info msg="initializing Dapr workflow component" app_id=proxy component="dapr (workflow.dapr/v1)" instance=d5ec20bb80c3 scope=dapr.contrib type=log ver=1.11.2
time="2023-08-02T16:09:05.190973985Z" level=debug msg="try to connect to placement service: dns:///localhost:50005" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.actor.internal.placement type=log ver=1.11.2
time="2023-08-02T16:09:05.191066825Z" level=info msg="dapr initialized. Status: Running. Init Elapsed 452ms" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime type=log ver=1.11.2
time="2023-08-02T16:09:05.325429179Z" level=debug msg="error connecting to placement service (will retry to connect in background): rpc error: code = Unavailable desc = last connection error: connection error: desc = \"transport: Error while dialing: dial tcp 127.0.0.1:50005: connect: connection refused\"" app_id=proxy instance=d5ec20bb80c3 scope=dapr.runtime.actor.internal.placement type=log ver=1.11.2
time="2023-08-02T16:09:34.739661875Z" level=debug msg="Refreshing all mDNS addresses." app_id=proxy component="mdns (nameResolution/v1)" instance=d5ec20bb80c3 scope=dapr.contrib type=log ver=1.11.2
time="2023-08-02T16:09:34.739798597Z" level=debug msg="no mDNS apps to refresh." app_id=proxy component="mdns (nameResolution/v1)" instance=d5ec20bb80c3 scope=dapr.contrib type=log ver=1.11.2
time="2023-08-02T16:10:04.739608871Z" level=debug msg="Refreshing all mDNS addresses." app_id=proxy component="mdns (nameResolution/v1)" instance=d5ec20bb80c3 scope=dapr.contrib type=log ver=1.11.2
time="2023-08-02T16:10:04.739688267Z" level=debug msg="no mDNS apps to refresh." app_id=proxy component="mdns (nameResolution/v1)" instance=d5ec20bb80c3 scope=dapr.contrib type=log ver=1.11.2
time="2023-08-02T16:10:34.73928827Z" level=debug msg="Refreshing all mDNS addresses." app_id=proxy component="mdns (nameResolution/v1)" instance=d5ec20bb80c3 scope=dapr.contrib type=log ver=1.11.2
time="2023-08-02T16:10:34.739394095Z" level=debug msg="no mDNS apps to refresh." app_id=proxy component="mdns (nameResolution/v1)" instance=d5ec20bb80c3 scope=dapr.contrib type=log ver=1.11.2

```

**Comment by ItalyPaleAle** [comment]
@ben-kotvis can you confirm, using for example jwt.io or jwt.ms, that:

- The value of `iss` in the token matches `https://login.microsoftonline.com/{tenant}/v2.0` (note this must use `tenant` as UUID/GUID and not as the name such as "contoso.onmicrosoft.com")
- The value of `aud` matches the audience, which is the client ID of the application (a UUID/GUID, bit a different one from the tenant)
- The token is not expired (`exp`), and the current time is not before `nbf` if present
- The server has the correct time, and there's no clock skew of more than 5 mins

**Comment by ben-kotvis** [comment]
> @ben-kotvis can you confirm, using for example jwt.io or jwt.ms, that:
> 
> * The value of `iss` in the token matches `https://login.microsoftonline.com/{tenant}/v2.0` (note this must use `tenant` as UUID/GUID and not as the name such as "contoso.onmicrosoft.com")
> * The value of `aud` matches the audience, which is the client ID of the application (a UUID/GUID, bit a different one from the tenant)
> * The token is not expired (`exp`), and the current time is not before `nbf` if present
> * The server has the correct time, and there's no clock skew of more than 5 mins

I can confirm that.  Just to re-iterate, I can take the exact same settings and token and it will work with 1.9.0 (although the metadate names have to be changed to issuerURL and clientID) as that was changed between those versions.

**Comment by ItalyPaleAle** [comment]
I can sort of reproduce some odd behavior here.

How are you obtaining the tokens?

I noticed that when obtaining it via the client credentials flow, I need too set a scope for the Microsoft Graph, so the audience and issuer are different.

**Comment by ItalyPaleAle** [comment]
Although I can't repro this issue correctly, I think this _may_ fix the bug: https://github.com/dapr/components-contrib/pull/3033

**Comment by ben-kotvis** [comment]
> Although I can't repro this issue correctly, I think this _may_ fix the bug: #3033

[This is the flow](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow)

This is an angular front end and the Bearer middleware is on the backend.  So I have to app registrations, one for the angular app and then I request a scope that's on the API.  This is pretty typical (in case it's not obvious).   I could probably create a VS code dev container for both scenarios if it helps.  Of course a lot of it has to do with the Azure AD setup.  But it's definitely a regression since 1.10.0.

**Comment by ItalyPaleAle** [comment]
Ok, if you are using the auth code flow then it makes sense. I think #3033 should fix the issue, as it fixed it in similar scenarios in the past.

**Comment by ItalyPaleAle** [comment]
@ben-kotvis the more I think about this, the more I'm feeling good about #3033 fixing this. I'm gonna close this for now, and maybe you can try with a nightly build (it will be a few days before the updated component gets merged into a nightly build)?

**Comment by ben-kotvis** [comment]
> @ben-kotvis the more I think about this, the more I'm feeling good about #3033 fixing this. I'm gonna close this for now, and maybe you can try with a nightly build (it will be a few days before the updated component gets merged into a nightly build)?

Sure, keep me posted and I will try it then.

**Comment by ItalyPaleAle** [comment]
@ben-kotvis we have nightly builds that contain the fix. Would it be possible for you to try them please and let us know if it works?
https://github.com/dapr/dapr/pkgs/container/dapr/116054466?tag=nightly-2023-08-05

**Comment by ben-kotvis** [comment]
@ItalyPaleAle is there a way to run this on the cli?  I initialize DAPR like this 

```dapr init --runtime-version 1.10.0```
